# Name Merge Test

#### This notebook is aimed at testing the name merge performance between different dataset:
1. Burning Glass merge to Shark Repellent (BG is the query list, SR is the search dictionary)
- Shark Repellent merge to Compustat (SR is the query list, Comp is the search dictionary)
- Burning Glass merge to Compustat (BG is the query list, Comp is the search dictionary)

In [1]:
import pandas as pd
import numpy as np
import jieba
import difflib
import random
from cleanco import cleanco
import jaro
import string

 #### Download Data

In [2]:
comp=pd.read_stata(r"..\Compustat\names.dta")
shark=pd.read_excel(r"..\FactSet SharkRepllent Data (Pulled 2019-11-19).xlsx").drop(index=[0,1,2,4])
shark.columns=shark.loc[3]
shark=shark.drop(index=3).reset_index().drop('index',axis=1)
bg=pd.read_stata(r"..\00_BGT_Firm_Names.dta")
#create name
c_name=pd.DataFrame(comp['conm']).astype(str)
s_name=pd.DataFrame(shark.iloc[:,7]).rename({'Company Name':'conm'}, axis=1)
bg_name=pd.DataFrame(bg.name_bgt[bg['total_postings_bgt']>50])
bg_name_full=pd.DataFrame(bg.name_bgt)

#### Provide cleaned versions of names
1. 
 `cleanco` processes company names, providing cleaned versions of the names by stripping away terms indicating organization type (such as "Ltd." or "Corp").  
- Using a database of organization type terms, It also provides an utility to deduce the type of organization, in terms of US/UK business entity types (ie. "limited liability company" or "non-profit"). 

- Details about this package can be found at https://pypi.org/project/cleanco/

- I also change uppercase letter to lowercase.


In [3]:
#clean name
#remove organization type and thansfer to lower case
s_name1={}.fromkeys(list(map(lambda x: cleanco(x.lower()).clean_name(), s_name.conm))).keys()
c_name1={}.fromkeys(list(map(lambda x: cleanco(x.lower()).clean_name(), c_name.conm))).keys()
bg_name1={}.fromkeys(list(map(lambda x: cleanco(x.lower()).clean_name(), bg_name.name_bgt))).keys()

In [4]:
def jaro_distance(list1,list2):
    """
    Measure strings similarity by Jaro-winker distance.
    
    Parameters
    ----------
        list1: list of query strings
        list2: list of names dictionary       
    Returns
    -------
        df: Dataframe with three columns: "query_name", "match" and "score"
        "query_name" is the query string(target company name)
        "match" is the most similary string found at dictionary for the query string
        "score" is a float number used to measure the similarity between the query string and "match" string, range(0,1)
              
    """
    df=pd.DataFrame(list1)
    label=[]
    score_get=np.empty(len(list1))
    score=np.empty(len(list2))
    for n1,str1 in enumerate(list1):
        for n2,str2 in enumerate(list2):
            score[n2]=jaro.jaro_winkler_metric(str1,str2)
        imax=np.argmax(score)
        label.append(list2[imax])
        score_get[n1]=max(score)
    df['match']=label
    df['score']=score_get
    df.rename(columns={0:'query_name'})
    return df

In [5]:
class StrSimilarity1():
    def __init__(self,word):
        self.word=word
#Compared函数，参数str_list是对比字符串列表
#返回原始字符串分词后和对比字符串的匹配次数，返回一个字典
    def Compared(self,str_list):
        """
        Count common words.
        
        Parameters
        ----------
        str_list: a list contains potential matched strings
        
        
        Returns
        ----------
        dict_data: a dictionary
        keys are the potential matched strings
        value are the number of common words
        
        """
        dict_data={}
        sarticiple=list(self.word.strip().translate(str.maketrans('', '', string.punctuation)).split())
        for strs in str_list:
            #s_name list
            strs_word=list(strs.strip().translate(str.maketrans('', '', string.punctuation)).split())
            num=0
            for strs1 in strs_word:
                if strs1 in sarticiple:
                    num = num+1
                else: 
                    num = num
            dict_data[strs]=num
        return dict_data
    #NumChecks函数，参数dict_data是原始字符串分词后和对比字符串的匹配次数的字典，也就是Compared函数的返回值
    #返回出现次数最高的两个，返回一个字典
    def NumChecks(self,dict_data):
        """
        Return two potential strings with the hightest common word number.
        
        
        Parameters
        ----------
        dict_data: a dictionary
        keys are the potential matched strings
        value are the number of common words(return of Compared)
        
        
        Returns
        ----------
        dict_data: a dictionary
        keys are the two potential matched strings with the highest number of common words
        value are the number of common words
        
        """       
        list_data = sorted(dict_data.items(), key=lambda asd:asd[1], reverse=True)
        length = len(list_data)
        json_data = {}
        if length>=2:
            datas = list_data[:2]
        else:
            datas = list_data[:length]
        for data in datas:
            json_data[data[0]]=data[1]
        return json_data
#MMedian函数，参数dict_data是出现次数最高的两个对比字符串的字典，也就是NumChecks函数的返回值
#返回对比字符串和调节值的字典
    def MMedian(self,dict_data):
        """
        Calculate adjusted similarity scores for most potential strings(optional step).
        
        
        Parameters
        ----------
        dict_data: a dictionary
        keys are the two potential matched strings with the highest number of common words
        value are the number of common words(return of NumChecks)
        
        
        Returns
        ----------
        dict_data: a dictionary
        keys are the two potential matched strings with the highest number of common words
        value are the adjusted similarity scores
               
        """   
        
        median_list={}
        l=len(list(self.word.strip().translate(str.maketrans('', '', string.punctuation)).split()))#query string word numbers
        for k,v in dict_data.items():#k is potential string, v is the common word number
            length=len(list(k.strip().translate(str.maketrans('', '', string.punctuation)).split()))#potential string word numbers
            if l>v: 
                if v==length:
                    xx=-1
                else: 
                    xx = ((abs(l-v))/l)
            else: 
                 xx=-2    
            median_list[k] = xx
        return median_list
    
    
    
#Appear函数，参数dict_data是对比字符串和调节值的字典，也就是MMedian函数的返回值
#返回最相似的字符串
    def Appear(self,dict_data):
        """
        Return the most similar potential string.
        
        
        Parameters
        ----------
        dict_data: a dictionary
        keys are the two potential matched strings with the highest number of common words
        value are the adjusted similarity scores(return of  MMedian)
        
        
        Returns
        ----------
        dict_data: a dictionary
        key is the query string
        value is most similar potential string
               
        """   
        json_data={}
        for k,v in dict_data.items():
            fraction = difflib.SequenceMatcher(None, self.word, k).quick_ratio()-v
            json_data[k]=fraction
        tulp_data = sorted(json_data.items(), key=lambda asd:asd[1], reverse=True)
        return tulp_data[0][0],tulp_data[0][1]   
    
def name_match1(query_list1,str_list1):
    """
    Measure strings similarity by StrSimilary.
    
    Parameters
    ----------
        query_list1: list of query strings
        str_list1: list of names dictionary       
    Returns
    -------
        df: Dataframe with three columns: "query_name", "match" and "score"
        "query_name" is the query string(target company name)
        "match" is the most similary string found at dictionary for the query string
        "score" is a float number used to measure the similarity between the query string and "match" string, range(0,1)             
    """ 
    name_match=[]
    score=[]
    #str_list1=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in str_list1)
    for i,str_query in enumerate(query_list1):
        def main():
            query_str =str_query
            str_list=str_list1
    
            ss = StrSimilarity1(query_str)
            list_data = ss.Compared(str_list)
            num = ss.NumChecks(list_data)
            mmedian = ss.MMedian(num)
            #print(query_str,ss.Appear(mmedian))
            return ss.Appear(mmedian)
        if __name__=="__main__":
            name_match.append(main()[0])
            score.append(main()[1])
    df=pd.DataFrame(query_list1)
    df['match']=name_match
    df['score']=score
    df.rename(columns={0:'query_name'})
    return df

#### 1. Burning Glass —>Shark Repellent

In [69]:
#real data test
random.seed(123)
query_bg=random.sample(bg_name1,100)#1000 names 
s_name1=sorted(s_name1, key=len)    
dict_s=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in s_name1)
query_bg=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in query_bg)
query_bg.sort(reverse=False)

#### 1.1 Merge Result

- StrSimilar

In [71]:
%%time
df_bg2s=name_match1(query_bg,dict_s).sort_values(by='score',ascending=False)

Wall time: 8.87 s


In [72]:
df_bg2s[df_bg2s.score>0]

0                         match  \
46                                it solutions          pomeroy it solutions   
97                                       triad               triad hospitals   
91                                 team sewell                          team   
66  northwestern connecticut community college                  northwestern   
75                rescare residential services                       rescare   
86                           spn well services              us well services   
37                          gm financial group          meta financial group   
59                         mb global logistics         echo global logistics   
36                    global commerce services      premiere global services   
82                       senior resource group            asa resource group   
39                hamadeh educational services      itt educational services   
94                          tpg hotels resorts          xenia hotels resorts   
52                      ledic management group  the management network group   
62                             mohegan holding                   gam holding   
32                            fewi development          bindview development   
43                           hub international             cai international   
2                              access security             strattec security   
16                          boulder scientific             medtox scientific   
72                          pinnacle marketing             pinnacle airlines   
34                              five star bank            five star products   
92                        thomas reprographics       thomas properties group   
20                         catapult technology                par technology   
80                                savory foods                   tyson foods   
47                            jefferson parish          jefferson bancshares   
33                              first business                  first united   
40                               hcr home care     national home health care   
79                              rrr automotive              sonic automotive   
42                              hotel granduca                nh hotel group   
85                           sovereign society             sovereign bancorp   
18         business technology managment group          art technology group   
17                                broken arrow           arrow international   
9                         amerifreight systems                   gse systems   
14                              bedard medical                   nmt medical   
95                                    tree com                   dollar tree   
0                                   1st access        1st century bancshares   
71                      paul sandhu industries                 sl industries   
50                                 lake county            lake shore bancorp   
87               spotsylvania county goverment   orange county business bank   
93                    timbar packaging display      graham packaging company   
15                       berlin public schools        urals energy public co   
35                   florida hospital memorial      florida public utilities   

       score  
46  2.750000  
97  2.500000  
91  1.533333  
66  1.444444  
75  1.400000  
86  0.575758  
37  0.561404  
59  0.516667  
36  0.500000  
82  0.487179  
39  0.474359  
94  0.456140  
52  0.386667  
62  0.346154  
32  0.333333  
43  0.323529  
2   0.312500  
16  0.300000  
72  0.300000  
34  0.291667  
92  0.290698  
20  0.287879  
80  0.282609  
47  0.277778  
33  0.269231  
40  0.245614  
79  0.233333  
42  0.214286  
85  0.205882  
18  0.190909  
17  0.145161  
9   0.145161  
14  0.140000  
95  0.131579  
0   0.125000  
71  0.076190  
50  0.051724  
87  0.047619  
93  0.041667  
15  0.031008  
35  0.027211

In [73]:
#threshold=1
df_bg2s[df_bg2s.score>1]

0                 match     score
46                                it solutions  pomeroy it solutions  2.750000
97                                       triad       triad hospitals  2.500000
91                                 team sewell                  team  1.533333
66  northwestern connecticut community college          northwestern  1.444444
75                rescare residential services               rescare  1.400000

- Jaro-winkler

In [76]:
%%time
df_bg2s_j=jaro_distance(query_bg,dict_s).sort_values(by='score',ascending=False)

Wall time: 33.7 s


In [79]:
#threshold=0.8
df_bg2s_j[df_bg2s_j.score>0.8]

0                        match     score
86             spn well services             us well services  0.939951
43             hub international           sfbc international  0.905229
47              jefferson parish         jefferson bancshares  0.900714
20           catapult technology         datapulse technology  0.895304
72            pinnacle marketing            pinnacle airlines  0.891690
..                           ...                          ...       ...
39  hamadeh educational services  hallmark financial services  0.814674
6      allfast fastening systems        alliance data systems  0.810593
29                       escreen                      rescare  0.809524
52        ledic management group              lee metal group  0.809062
30                        eshots                         hess  0.805556

[61 rows x 3 columns]

In [80]:
#threshold=0.8
df_jaro[df_jaro.score>0.85]

0  \
86                           spn well services   
43                           hub international   
47                            jefferson parish   
20                         catapult technology   
72                          pinnacle marketing   
33                              first business   
45                                 indian eyes   
99                                   watertown   
83                                      shogun   
97                                       triad   
73                                    pronexus   
91                                 team sewell   
82                       senior resource group   
61                                mobilenow pa   
51                               landi english   
46                                it solutions   
3    advanced dermatology and cosmetic surgery   
62                             mohegan holding   
85                           sovereign society   
8                americaninsuranceorganization   
36                    global commerce services   
7                                    altsource   
32                            fewi development   
92                        thomas reprographics   
37                          gm financial group   
58                           matheson trucking   
34                              five star bank   
66  northwestern connecticut community college   
9                         amerifreight systems   
65                                   noralogic   

                              match     score  
86                 us well services  0.939951  
43               sfbc international  0.905229  
47             jefferson bancshares  0.900714  
20             datapulse technology  0.895304  
72                pinnacle airlines  0.891690  
33                     first united  0.889524  
45                indiana resources  0.879465  
99                           waters  0.877778  
83                            shofu  0.875556  
97                           trimas  0.875556  
73                           prodex  0.873611  
91                             team  0.872727  
82  senior housing properties trust  0.869508  
61                       mobileiron  0.868333  
51               landcadia holdings  0.867949  
46                    pdf solutions  0.867521  
3         advanced drainage systems  0.867407  
62                macmahon holdings  0.866335  
85                sovereign bancorp  0.865686  
8      american international group  0.863412  
36                   global sources  0.863278  
7            altisource residential  0.862121  
32                grace development  0.859069  
92          thomas properties group  0.859003  
37                     gs financial  0.858333  
58                         mathstar  0.857353  
34           five star quality care  0.857143  
66                     northwestern  0.857143  
9                       americredit  0.853636  
65                        corelogic  0.851852

In [83]:
#threshold=0.8
df_jaro[df_jaro.score>0.87]

0                 match     score
86    spn well services      us well services  0.939951
43    hub international    sfbc international  0.905229
47     jefferson parish  jefferson bancshares  0.900714
20  catapult technology  datapulse technology  0.895304
72   pinnacle marketing     pinnacle airlines  0.891690
33       first business          first united  0.889524
45          indian eyes     indiana resources  0.879465
99            watertown                waters  0.877778
83               shogun                 shofu  0.875556
97                triad                trimas  0.875556
73             pronexus                prodex  0.873611
91          team sewell                  team  0.872727

#### 1.2  Merge Performance Evaluation
- According to the test result, the threshold for `StrSimilarity` should be larger than 0.6; the threshold for `jaro-winkler` should be larger than 0.89;
- Larger score in `StrSimilarity` provides a better sensible match, which means that we can find a clear cut-off for scores;
- For `jaro-winkler`, a larger score does not means a better sensible match: "northwestern connecticut community college" and	"northwestern" with score 0.857143, "indian eyes" and	"indiana resources" woth score 0.879465. However, the former match is more likely to be correct than the latter case;
- So the `StrSimilarity`is much stable than `jaro-winkler`

------------------------------------------------------------

#### 2. Shark Repellent —>Compustat

In [21]:
#real data test
# ccompustat is dictionary;shark repellent is query name
random.seed(123)
query_s=random.sample(s_name1,100)#1000 names 
c_name1=sorted(c_name1, key=len)    
dict_c=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in c_name1)#remove punctuation
query_s=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in query_s)
query_s.sort(reverse=False)

#### 2.1 Merge Result

- StrSimilar

In [24]:
%%time
df_s2c=name_match1(query_s,dict_c).sort_values(by='score',ascending=False)
#take longer time since the dictionary is larger

Wall time: 1min


In [33]:
df_s2c[df_s2c.score>0.5]#the propert threshold should be 0.6

0                        match     score
0                              abb                          abb  3.000000
63                qts realty trust             qts realty trust  3.000000
38                juniper networks             juniper networks  3.000000
42                lancaster colony             lancaster colony  3.000000
43               life time fitness            life time fitness  3.000000
..                             ...                          ...       ...
48   mfs california municipal fund  mfs california municipal fd  0.714286
23  clear channel outdoor holdings  clear channel outdoor hldgs  0.697368
93                 us home systems             u s home systems  0.634409
20        centuria industrial reit        dream industrial reit  0.511111
69              rny property trust         arbor property trust  0.508772

[69 rows x 3 columns]

In [60]:
df_s2c[(df_s2c.score>0.6) & (df_s2c.score<2)]

0  \
85                           the lamson sessions   
86                             the malaysia fund   
46                                    mb bancorp   
84           the gabelli global multimedia trust   
53  nuveen florida quality income municipal fund   
73                silicon graphics international   
64                              quantum graphite   
48                 mfs california municipal fund   
23                clear channel outdoor holdings   
93                               us home systems   

                           match     score  
85               lamson sessions  1.882353  
86                 malaysia fund  1.866667  
46                       bancorp  1.823529  
84      gabelli multimedia trust  1.813559  
53  nuveen municipal income fund  1.777778  
73              silicon graphics  1.695652  
64                       quantum  1.608696  
48   mfs california municipal fd  0.714286  
23   clear channel outdoor hldgs  0.697368  
93              u s home systems  0.634409

- Jaro-winkler

In [34]:
%%time
df_s2c_j=jaro_distance(query_s,dict_c).sort_values(by='score',ascending=False)

Wall time: 3min 51s


In [86]:
df_s2c_j[df_s2c_j.score>0.5]#the propert threshold should be 0.7424

0                        match     score
0                                   abb                          abb  1.000000
63                     qts realty trust             qts realty trust  1.000000
38                     juniper networks             juniper networks  1.000000
42                     lancaster colony             lancaster colony  1.000000
43                    life time fitness            life time fitness  1.000000
..                                  ...                          ...       ...
15          bond international software              bond industries  0.832222
55    opportunity investment management  pzena investment management  0.774832
84  the gabelli global multimedia trust     gabelli multimedia trust  0.742460
77                     société générale                  ast general  0.742424
41                                  l90             delta 9 cannabis  0.597222

[100 rows x 3 columns]

In [55]:
df_s2c_j[(df_s2c_j.score>0.7) & (df_s2c_j.score<0.9)]

0  \
68                                        rision   
40                        koninklijke vendex kbb   
58                                 panoro energy   
71                security devices international   
81                          tegel group holdings   
74                                         sirva   
35               institutional financial markets   
94                                      usellcom   
16                          butler manufacturing   
83                         telegraaf media groep   
20                      centuria industrial reit   
53  nuveen florida quality income municipal fund   
97                                 worleyparsons   
39                        katakura industries co   
75                                           sjw   
7                                         amsurg   
78                                   speedy hire   
85                           the lamson sessions   
69                            rny property trust   
86                             the malaysia fund   
99                             zenyatta ventures   
15                   bond international software   
55             opportunity investment management   
84           the gabelli global multimedia trust   
77                              société générale   

                           match     score  
68                     radvision  0.900000  
40               koninklijke kpn  0.899654  
58             pan orient energy  0.893979  
71    security environmental sys  0.886778  
81        thistle group holdings  0.882121  
74                         shiva  0.880000  
35    institutional investorsdel  0.877975  
94                         selco  0.875000  
16                    butler mfg  0.870000  
83                       telemed  0.866667  
20            centurion minerals  0.865000  
53  nuveen fl quality inc mun fd  0.862987  
97                    worldspace  0.858077  
39                urt industries  0.854978  
75                     sjw group  0.844444  
7                         amstar  0.844444  
78                  speedfamipec  0.843939  
85               lamson sessions  0.840936  
69       republic property trust  0.837862  
86             thai capital fund  0.833484  
99              stealth ventures  0.833428  
15               bond industries  0.832222  
55   pzena investment management  0.774832  
84      gabelli multimedia trust  0.742460  
77                   ast general  0.742424

#### 2.2  Merge Performance Evaluation
- According to the test result, the threshold for `StrSimilarity` should be larger than 0.5 (maybe around 0.6344); the threshold for `jaro-winkler` should be larger than 0.74;
- Larger score in `StrSimilarity` provides a better sensible match, which means that we can find a clear cut-off for scores;
- For `jaro-winkler`, a larger score does not means a better sensible match: "the gabelli global multimedia trust" and	"gabelli multimedia trust" with score 0.742460, "zenyatta ventures" and "ventures" woth score 0.833428. However,the former match is more likely to be correct than the latter case;
- So the `StrSimilarity`is much stable than `jaro-winkler`

#### 3. Burning Glass —>Compustat

In [35]:
#real data test
# ccompustat is dictionary;shark repellent is query name
random.seed(123)
query_bg=random.sample(bg_name1,100)#1000 names 
c_name1=sorted(c_name1, key=len)    
dict_c=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in c_name1)#remove punctuation
query_bg=list(' '.join(str1.strip().translate(str.maketrans('', '', string.punctuation)).split()) for str1 in query_bg)
query_bg.sort(reverse=False)

#### 3.1 Merge Result

- StrSimilar

In [61]:
%%time
df_bg2c=name_match1(query_bg,dict_c).sort_values(by='score',ascending=False)
#take longer time since the dictionary is larger

Wall time: 1min


In [62]:
df_bg2c[df_bg2c.score>0.5]#the propert threshold should be 1

0                    match  \
43                           hub international        hub international   
33                              first business   business first bancshs   
46                                it solutions     pomeroy it solutions   
68                                       orbus             orbus pharma   
97                                       triad            triad systems   
0                                   1st access                   access   
47                            jefferson parish                jefferson   
13                               banco popular                  popular   
2                              access security                   access   
91                                 team sewell                     team   
84        smithsonian national zoological park            park national   
60                              medstaff local                    local   
66  northwestern connecticut community college             northwestern   
75                rescare residential services                  rescare   
77                    rogers behavioral health                   rogers   
69               oregon community credit union                    union   
86                           spn well services         us well services   
37                          gm financial group       kb financial group   
52                      ledic management group    icis management group   
94                          tpg hotels resorts      host hotels resorts   
82                       senior resource group  business resource group   
59                         mb global logistics    echo global logistics   

       score  
43  3.000000  
33  2.777778  
46  2.750000  
68  2.588235  
97  2.555556  
0   1.750000  
47  1.720000  
13  1.700000  
2   1.571429  
91  1.533333  
84  1.530612  
60  1.526316  
66  1.444444  
75  1.400000  
77  1.400000  
69  1.294118  
86  0.575758  
37  0.555556  
52  0.550388  
94  0.531532  
82  0.530303  
59  0.516667

- Jaro Winkler

In [63]:
%%time
df_bg2c_j=jaro_distance(query_bg,dict_c).sort_values(by='score',ascending=False)

Wall time: 4min 5s


In [66]:
df_bg2c_j[df_bg2c_j.score>0.8]

0                       match     score
43              hub international           hub international  1.000000
97                          triad                       traid  0.946667
37             gm financial group         bmr financial group  0.946394
86              spn well services            us well services  0.939951
27                            ebc                        enbc  0.925000
..                            ...                         ...       ...
10                         arnamy                     aradyme  0.826032
41  hillcrest property management  hillcrest resources ltdold  0.824481
57    massage heights franchising          mass merchandisers  0.824444
23                csu bakersfield                  clearfield  0.816667
87  spotsylvania county goverment  sports entertainment group  0.812983

[88 rows x 3 columns]

In [87]:
df_bg2c_j[df_bg2c_j.score>0.85]

0  \
43                              hub international   
97                                          triad   
37                             gm financial group   
86                              spn well services   
27                                            ebc   
33                                 first business   
46                                   it solutions   
40                                  hcr home care   
12                                          avivo   
2                                 access security   
7                                       altsource   
32                               fewi development   
47                               jefferson parish   
85                              sovereign society   
95                                       tree com   
68                                          orbus   
72                             pinnacle marketing   
73                                       pronexus   
45                                    indian eyes   
55                                    logicsource   
38                           goldbergs fine foods   
36                       global commerce services   
61                                   mobilenow pa   
78                                           rokt   
88                       standard plumbing supply   
13                                  banco popular   
35                      florida hospital memorial   
66     northwestern connecticut community college   
54                             life bridge health   
99                                      watertown   
8                   americaninsuranceorganization   
92                           thomas reprographics   
91                                    team sewell   
52                         ledic management group   
34                                 five star bank   
79                                 rrr automotive   
94                             tpg hotels resorts   
51                                  landi english   
58                              matheson trucking   
26                         detroit institute arts   
3       advanced dermatology and cosmetic surgery   
18            business technology managment group   
20                            catapult technology   
71                         paul sandhu industries   
9                            amerifreight systems   
21                          chicago pizza brewery   
5                           all seasons placement   
62                                mohegan holding   
17                                   broken arrow   
89                                stephens little   
64                                    nomadhealth   
30                                         eshots   
28                                     employindy   
16                             boulder scientific   
19  carolinas continuecare hospital at university   
77                       rogers behavioral health   
65                                      noralogic   

                           match     score  
43             hub international  1.000000  
97                         traid  0.946667  
37           bmr financial group  0.946394  
86              us well services  0.939951  
27                          enbc  0.925000  
33                   first busey  0.924675  
46                 viq solutions  0.920940  
40                  thc homecare  0.920940  
12                         aviva  0.920000  
2                      accessity  0.920000  
7                   healthsource  0.916667  
32                ec development  0.913690  
47                     jefferson  0.912500  
85      sovereign specialty chem  0.909069  
95                       trex co  0.908333  
68                         orbis  0.906667  
72             pinnacle airlines  0.891690  
73                        pronet  0.891667  
45                   indian head  0.890909  
55                         logic  0.890909  
38            golden state f

#### 3.2  Merge Performance Evaluation
- According to the test result, the threshold for `StrSimilarity` should be larger than 0.5757 (maybe around 0.6344); the threshold for `jaro-winkler` should be larger than 0.86;
- Larger score in `StrSimilarity` provides a better sensible match, which means that we can find a clear cut-off for scores;
- For `jaro-winkler`, a larger score does not means a better sensible match: "team sewell" and	"team" with score 0.872727, "pronexus" and "pronet" woth score 0.891667.However, the former match is more likely to be correct than the latter case;
- So the `StrSimilarity`is much stable than `jaro-winkler`  

### Conclusion
1. `StrSimilarity`'s thresholds are around 0.6 across the three dataset, while the threshold for `Jaro-winkler`'s vary a lot for different data sets. 
- `StrSimilarity`'s score can be regarded as a monotone function of the probability of correct match, while `Jaro-winkler`'s not.